Handle HTML entities in textual data

zmwangx opened the pull request:
Fixes the issue where single quotes and double quotes are stripped from output (both result titles and result texts).

Before:

```
> googler -C double quotes </dev/null
 1 Quotation mark - Wikipedia, the free encyclopedia 
https://en.wikipedia.org/wiki/Quotation_mark
In American English, double quotes are used normally (the primary style). If quote marks are used inside another pair 
of quote marks, then single quotes are... 

 2 Single quotes or double quotes? Its really quite simple. - Slate 
http://www.slate.com/blogs/lexicon_valley/2014/10/21/single_quotes_or_double_quotes_it_s_really_quite_simple.html
Oct 21, 2014 - For several years now in teaching writing classes to college freshmen, Ive noticed some students adopt 
another rule: double quotes for long... 

 3 Quotation Marks: When to Use Double or Single Quotation ... 
https://www.scribendi.com/advice/when_to_use_double_or_single_quotation_marks.en.html
Quotations within the block will have double or single quotes, according to the convention being used (British or 
American). As usual, these different conventions... 

 4 Double Quotes - BrainyQuote 
http://www.brainyquote.com/quotes/keywords/double.html
Double Quotes from BrainyQuote, an extensive collection of quotations by famous authors, celebrities, and newsmakers. 

 5 Grammar Girl : Single Quotation Marks Versus Double ... 
http://www.quickanddirtytips.com/education/grammar/single-quotation-marks-versus-double-quotation-marks
Aug 18, 2011 - Double Quotation Marks for Scare Quotes. Double quotation marks can also be used sometimes to indicate 
that a word is special in some way. 

 6 When do you use double quotation marks? - APA Style 
http://www.apastyle.org/learn/faqs/use-double-quotes.aspx
Provides APA Style guidelines on when to use double quotation marks. 

 7 Quotation Marks - Capital Community College 
http://grammar.ccc.commnet.edu/grammar/marks/quotation.htm
My mothers favorite quote was from Shakespeare: This above all, to thine own .... In fact, single-quote marks and 
double-quote marks are apt to be reversed in... 

 8 Straight and curly quotes | Buttericks Practical Typography 
http://practicaltypography.com/straight-and-curly-quotes.html
Straight quotes are the two generic vertical quotation marks located near the return key: the straight single quote ( 
) and the straight double quote ( ). 

 9 Bash Reference Manual: Double Quotes - GNU 
https://www.gnu.org/s/bash/manual/html_node/Double-Quotes.html
Enclosing characters in double quotes ( ) preserves the literal value of all characters within the quotes, with the 
exception of $ , ` , \ , and, when history... 

Enter 'n', 'p', 'g keywords' or result number to continue: 
```

After:

```
> googler -C double quotes </dev/null
 1 Quotation mark - Wikipedia, the free encyclopedia 
https://en.wikipedia.org/wiki/Quotation_mark
In American English, double quotes are used normally (the "primary" style). If quote marks are used inside another 
pair of quote marks, then single quotes are ... 

 2 Single quotes or double quotes? It's really quite simple. - Slate 
http://www.slate.com/blogs/lexicon_valley/2014/10/21/single_quotes_or_double_quotes_it_s_really_quite_simple.html
Oct 21, 2014 - For several years now in teaching writing classes to college freshmen, I've noticed some students 
adopt another rule: double quotes for long ... 

 3 Quotation Marks: When to Use Double or Single Quotation ...›› 
https://www.scribendi.com/advice/when_to_use_double_or_single_quotation_marks.en.html
Quotations within the block will have double or single quotes, according to the convention being used (British or 
American). As usual, these different conventions ... 

 4 Double Quotes - BrainyQuote 
http://www.brainyquote.com/quotes/keywords/double.html
Double Quotes from BrainyQuote, an extensive collection of quotations by famous authors, celebrities, and newsmakers. 

 5 Grammar Girl : Single Quotation Marks Versus Double ... 
http://www.quickanddirtytips.com/education/grammar/single-quotation-marks-versus-double-quotation-marks
Aug 18, 2011 - Double Quotation Marks for Scare Quotes. Double quotation marks can also be used sometimes to indicate 
that a word is special in some way. 

 6 When do you use double quotation marks? - APA Style 
http://www.apastyle.org/learn/faqs/use-double-quotes.aspx
Provides APA Style guidelines on when to use double quotation marks. 

 7 Quotation Marks - Capital Community College 
http://grammar.ccc.commnet.edu/grammar/marks/quotation.htm
My mother's favorite quote was from Shakespeare: "This above all, to thine own .... In fact, single-quote marks and 
double-quote marks are apt to be reversed in ... 

 8 Straight and curly quotes | Butterick's Practical Typography 
http://practicaltypography.com/straight-and-curly-quotes.html
Straight quotes are the two generic vertical quotation marks located near the return key: the straight single quote ( 
' ) and the straight double quote ( " ). 

 9 Bash Reference Manual: Double Quotes - GNU 
https://www.gnu.org/s/bash/manual/html_node/Double-Quotes.html
Enclosing characters in double quotes (' " ') preserves the literal value of all characters within the quotes, with 
the exception of ' $ ', ' ` ', ' \ ', and, when history ... 

Enter 'n', 'p', 'g keywords' or result number to continue: 
```

Marked improvement! Although the code behind is really dirty.

jarun commented:
A quick summary of the processing overhead?

zmwangx commented:
Well, the additional work is O(n) in the total number of HTML entities. Negligible.

zmwangx commented:
And the work cannot be avoided if you want to capture the quotes at all.

jarun commented:
I understand. I was just afraid if we are going to parse the complete html
data first.


Regards,

Arun

http://tuxdiary.com/

On Wed, Mar 2, 2016 at 2:22 PM, Zhiming Wang <notifications@github.com>
wrote:

> And the work cannot be avoided if you want to capture the quotes at all.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jarun/googler/pull/39#issuecomment-191133485>.
>

zmwangx commented:
> I was just afraid if we are going to parse the complete html data first.

Rest assured, we don't. We are operating in the framework of HTMLParser, in a very standard way.

>  the additional work is O(n) in the total number of HTML entities.

Sorry that's not true, because we have to assign the handlers, which is linear in the number of h3 tags and effective text nodes combined. Still, this overhead could not be avoided since it's basically imposed by HTMLParser.

jarun commented:
> Sorry that's not true, because we also have to assign the handlers, which is linear in the number of h3 tags and effective text nodes combined. Still, this overhead could not be avoided since it's basically imposed by HTMLParser.

Thought so, but we can live it. I have also added a few optimizations yesterday. Hope these negate each other :).

zmwangx commented:
Just realized I wasn't properly unregistering handlers, resulting in weird characters in the top post:

```
 3 Quotation Marks: When to Use Double or Single Quotation ...›› 
```

Should have been fixed. Changes since last version:

```diff
diff --git a/googler b/googler
index 85c35b3..d65d6de 100755
--- a/googler
+++ b/googler
@@ -149,6 +149,7 @@ class GoogleParser(HTMLParser.HTMLParser):
         if tag == "h3":
             self.handle_starttag = self.div_outer_start
             self.handle_data = self.div_outer_data
+            self.unset_entityref_and_charref_handlers()
             self.handle_endtag = self.div_outer_end
 
     # inner <div> ... </div>
@@ -179,6 +180,7 @@ class GoogleParser(HTMLParser.HTMLParser):
         if tag == "span":
             self.handle_starttag = self.div_outer_start
             self.handle_data = self.div_outer_data
+            self.unset_entityref_and_charref_handlers()
             self.handle_endtag = self.div_outer_end
 
     def span_inner_start(self, tag, start):
@@ -191,6 +193,7 @@ class GoogleParser(HTMLParser.HTMLParser):
         if tag == "span":
             self.handle_starttag = self.span_outer_start
             self.handle_data = self.span_outer_data
+            self.unset_entityref_and_charref_handlers()
             self.handle_endtag = self.span_outer_end
 
     # Set entityref and charref handlers so that decoded refs are
@@ -200,6 +203,9 @@ class GoogleParser(HTMLParser.HTMLParser):
         self.handle_entityref = lambda ref: self.entityref(dest, ref)
         self.handle_charref = lambda ref: self.charref(dest, ref)
 
+    def unset_entityref_and_charref_handlers(self):
+        self.handle_entityref = self.handle_charref = lambda ref: None
+
     def entityref(self, dest, ref):
         try:
             char = unichr(name2codepoint[ref])
```

zmwangx commented:
Seems to have introduced another problem:

```
 2  Single quotes or double quotes? It's really quite simple. - Slate
http://www.slate.com/blogs/lexicon_valley/2014/10/21/single_quotes_or_double_quotes_it_s_really_quite_simple.html
Oct 21, 2014 - For several years now in teaching writing classes to college freshmen, Ive noticed some students adopt
another rule: double quotes for long...
```

Note the missing single quote in `Ive`. Hell, need to invest more time to sort out the logical structure.

jarun commented:
Please take you time.

jarun commented:
The 'single' quotes are different. Check this [enlarged pic](http://i.imgur.com/FByGb99.png) taken from two consecutive lines on the page.

zmwangx commented:
> The 'single' quotes are different.

Not really. Here's what I got:

<img width="520" alt="screen shot 2016-03-02 at 10 58 00 am" src="https://cloud.githubusercontent.com/assets/4149852/13471511/bb238870-e065-11e5-87da-abbc548481de.png">

Notice the straight quote. More definitively, here's the excerpt from the response:

```html
</div><span class="st"><span class="f">Oct 21, 2014 - </span>For several years now in teaching writing classes to college freshmen, I&#39;ve noticed some students adopt another rule: <em>double quotes</em> for long&nbsp;...</span>
```

`&#39;` is U+0027. So either we are receiving different responses, or your editor performed some sort of smart quote shenanigans...

zmwangx commented:
Hopefully I've done it right this time.

jarun commented:
Thank you! Are you still testing or shall I merge it?

zmwangx commented:
> Are you still testing or shall I merge it?

It should be all good. I'll rebase and remove the do not merge and WIP tags.

> Also, just for my knowledge, is charref and entityref the standard way this is done in Python (2.x and 3.x)? 

I believe so. I briefly looked at BeautifulSoup's code, and it's basically done in the same way.

jarun commented:
Thanks for confirming.

zmwangx commented:
Squashed, good to go.

jarun commented:
Merged. Thank you!
